pulp_ansible/app/tasks/signature.py

```diff
@@ -0,0 +1,91 @@
+import asyncio
+import os
+import json
+import tempfile
+
+from pulpcore.plugin.stages import (
+    ArtifactSaver,
+    ContentSaver,
+    DeclarativeArtifact,
+    DeclarativeContent,
+    DeclarativeVersion,
+    Stage,
+    QueryExistingArtifacts,
+    QueryExistingContents,
+)
+
+from pulp_ansible.app.models import (
+    AnsibleRepository,
+    CollectionVersion,
+    CollectionVersionSignature,
+)
+
+from pulpcore.plugin.models import Artifact, SigningService
+from pulpcore.plugin.sync import sync_to_async_iterable
+
+
+def sign(repository_href, content_hrefs, signing_service_href):
+    """The signing task."""
+    repository = AnsibleRepository.objects.get(pk=repository_href)
+    if content_hrefs == ["*"]:
+        content = repository.latest_version().content.filter(pulp_type=CollectionVersion.get_pulp_type())
+    else:
+        content = CollectionVersion.objects.filter(pk__in=content_hrefs)
+    signing_service = SigningService.objects.get(pk=signing_service_href)
+    first_stage = CollectionSigningStage(content, signing_service)
+    SigningDeclarativeVersion(first_stage, repository).create()
+
+
+class SigningDeclarativeVersion(DeclarativeVersion):
+    """Custom signature pipeline."""
+
+    def pipeline_stages(self, new_version):
+        """The stages for the signing process."""
+        pipeline = [
+            self.first_stage,  # CollectionSigningStage
+            QueryExistingArtifacts(),
+            ArtifactSaver(),
+            # QueryExistingContents(),
+            ContentSaver(),
+        ]
+        return pipeline
+
+
+class CollectionSigningStage(Stage):
+    """
+    This stage signs the content and creates CollectionVersionSignatures.
+    """
+
+    def __init__(self, content, signing_service):
+        super().__init__()
+        self.content = content
+        self.signing_service = signing_service
+        self.base_dir = os.getcwd()
+
+    async def run(self):
+        """does the magic."""
+        async for collection_version in sync_to_async_iterable(self.content.iterator()):
```

Comment:
What if we filtered this some here to exclude those "already signed CollectionVersions" ? Perhaps my suggestion is not a good one because we like to let the stages do the work.

This gets me thinking about a problematic case though if all we can do is sign everything in a repo, then repo version has collectionversion A-1.1.1, and it gets signed. Then B-2.2.2 gets added and signed it will sign A-1.1.1 again. And then C-3.3.3 gets added it will sign A-1.1.1 and B-2.2.2 again and so forth.

After thinking this over more, I think we should have it generate signatures just for CollectionVersion items that do not have signatures.

Reply:
The API also takes an optional list of content to sign so you don't have to sign everything in the repository. I was thinking of making this optional list also be added to the repository if not already present, but maybe that feature would be too much functionality.

As for an exclude "already signed content", I think that would be a good idea for use cases of syncing new content and then signing them with your own keys. I think users might want to sign content with two different services, so this should be an option and not the only default.